google-chart-loader.html

```diff
@@ -0,0 +1,206 @@
+    },
+    'scatter': {
+      ctor: 'ScatterChart',
+    },
+    'stepped-area': {
+      ctor: 'SteppedAreaChart',
+    },
+    'table': {
+      ctor: 'Table',
+      pkg: 'table',
+    },
+    'timeline': {
+      ctor: 'Timeline',
+      pkg: 'timeline',
+    },
+    'gauge': {
+      ctor: 'Gauge',
+      pkg: 'gauge',
+    },
+    'treemap': {
+      ctor: 'TreeMap',
+      pkg: 'treemap',
+    },
+  };
+
+  function namespaceForType(type) {
+    return google[type.startsWith('md-') ? 'charts' : 'visualization'];
+  }
+
+  var packagesToLoad = {};
+  var promises = {};
+  var resolves = {};
+
+  Polymer({
+    is: 'google-chart-loader',
+    properties: {
+      /**
+       * Adds packages to the list of packages to load.
+       *
+       * This is an array consisting of any Google Visualization package names.
+       *
+       */
+      prime: {
+        type: Array,
+        value: function() { return []; },
+        observer: '_load',
+      },
+      /**
+       * Loads the package for the chart type specified.
+       *
+       * This may be any of the supported `google-chart` types.
+       * This is mainly used by the `google-chart` element internally.
+       *
+       */
+      type: {
+        type: String,
+        observer: '_loadForType',
+      },
+    },
+
+    get _corePackage() {
+      if (promises[DEFACTO_CHART_PACKAGE]) {
+        return promises[DEFACTO_CHART_PACKAGE];
+      }
+      return this._load([DEFACTO_CHART_PACKAGE]).then(function(pkgs) {
+        return pkgs[0];
+      });
+    },
+
+    _loadPackages: function() {
+      this.debounce('loadPackages', function() {
+        var packages = Object.keys(packagesToLoad);
+        if (!packages.length) {
+          return;
+        }
+        packagesToLoad = {};
+        google.charts.load('current', {'packages': packages});
+        google.charts.setOnLoadCallback(function() {
+          packages.forEach(function(pkg) {
+            this.fire('loaded', pkg);
+            resolves[pkg](google.visualization);
+          }.bind(this));
+        }.bind(this));
+      }, 100);
+    },
+
+    _load: function(pkgs) {
+      var returns = [];
+      pkgs.forEach(function(pkg) {
+        if (!promises[pkg]) {
+          packagesToLoad[pkg] = true;
+          promises[pkg] = new Promise(function(resolve) {
+            resolves[pkg] = resolve;
+          });
+          this._loadPackages();
+        }
+        returns.push(promises[pkg]);
+      }.bind(this));
+      return Promise.all(returns);
+    },
+
+    _loadForType: function(type) {
+      var chartData = CHART_CONSTRUCTORS[type];
+      return this._load([chartData.pkg || DEFACTO_CHART_PACKAGE])
+        .then(function() {
+          return namespaceForType(type)[chartData.ctor];
+        });
+    },
+
+    create: function(type, el) {
+      return this._loadForType(type).then(function(ctor) {
+        return new ctor(el);
+      });
+    },
+
+    fireOn: function(chart, eventName, opt_once) {
```

Comment:
Hmm... I'm definitely not sold on the name `fireOn`, either, but I think fireChartsEvent implies "now". I actually started with the name `subscribeTo` and `listenTo`. The difference is that we're not actually attaching any listeners at the caller level, we're just telling it to do a Poylmer `fire` on a chart event. That's why I went with `fireOn`.

What about `fireOnChartEvent`?